NugetLightRequest.cs

```diff
@@ -1909,5 +1930,169 @@ private IEnumerable<PackageItem> SearchForPackages(PackageSource source, string
+            if (credProviderPath.IsNullOrEmpty() && osPlatform != PlatformID.Unix)
+            {
+                string vswhereExePath = "";
+                // Check both Program Files x86, and Program Files, respectively
+                string vswhereExePath64 = "%ProgramFiles(x86)%\\Microsoft Visual Studio\\Installer\\vswhere.exe";
+                string vswhereExePath32 = "%ProgramFiles%\\Microsoft Visual Studio\\Installer\\vswhere.exe";
+                string fullVswhereExePath64 = Environment.ExpandEnvironmentVariables(vswhereExePath64);
+                string fullVswhereExePath32 = Environment.ExpandEnvironmentVariables(vswhereExePath32);
+                if (File.Exists(vswhereExePath64))
+                {
+                    // If the env variable exists, check to see if the path itself exists
+                    vswhereExePath = fullVswhereExePath64;
+                }
+                else if (File.Exists(vswhereExePath32))
+                {
+                    vswhereExePath = fullVswhereExePath32;
+                }
+
+                // Using a process to run VsWhere.exe so that we can find the installation path of Visual Studio
+                Process process = new Process();
+                process.StartInfo.FileName = vswhereExePath;
+                process.StartInfo.RedirectStandardOutput = true;
+                process.StartInfo.CreateNoWindow = false;
+                process.StartInfo.UseShellExecute = false;
+                string vsInstallationPath = "";
+                try
+                {
+                    process.Start();
+                    StreamReader reader = process.StandardOutput;
+
+                    while (!reader.EndOfStream)
+                    {
+                        string line = reader.ReadLine();
+                        if (Regex.IsMatch(line, @"installationPath"))
+                        {
+                            // Match all chars after 'installationPath:'
+                            Match vsInstallPathMatch = Regex.Match(line, @"(?<=\: ).*");
+                            vsInstallationPath = vsInstallPathMatch.ToString();
+                            break;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine(e.Message);
+                }
+                finally
+                {
+                    process.Close();
+                }
+                // Then use the installation path discovered by vswhere.exe to create the path to search for credential provider
+                // ex: "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise" + "\Common7\IDE\CommonExtensions\Microsoft\NuGet\Plugins\CredentialProvider.Microsoft\CredentialProvider.Microsoft.exe"
+                if (vsInstallationPath.IsNullOrEmpty())
+                {
+                    throw new ArgumentNullException("vsInstallationPath");
```

Comment:
argumentnullexception is for when someone passes a null into your function when you don't expect it. I'd use InvalidOperationException here. Also consider: what's the contract for this function? Is it 'best effort', ie return credentials if you can, return null if not, or should we throw exceptions to indicate there's a problem?

Reply:
I started off thinking of it in terms of return credentials if you can, return null if not.  Then started including some exception handlings.  Tbh, I think the former makes more sense in this context.